### DISCUSSION TITLE
Testing JetStream Ack/Nak

### DISCUSSION THREAD

**Discussion by paulomf** [nats-io/nats.net#263] [Ideas]
I've a method that looks something like this:
```
public async Task ProcessMessage(NatsJSMsg<NatsMemoryOwner<byte>> msg) {
    var success = Process(msg); // Some logic...
    if(success) {
        await msg.AckAsync();
    } else {
        await msg.NakAsync();
    }
}
```

Given a JetStream message, It executes some logic and based on the processing it'll ack or nak.

I'm trying unit test that method, however it has proven to be a challenge given that NatsJSMsg's constructor is internal and does not implement any interface that could be mocked.

Here's what I've came up with to validate if a message has been acked.

```
    [Test]
    public async Task Test_Ack() {
        var connection = Substitute.For<INatsConnection>();

        var msg = (NatsJSMsg<NatsMemoryOwner<byte>>)Activator.CreateInstance(
            typeof(NatsJSMsg<NatsMemoryOwner<byte>>),
            BindingFlags.Instance | BindingFlags.NonPublic,
            null,
            new object[] {
                new NatsMsg<NatsMemoryOwner<byte>> {
                    Data = NatsMemoryOwner<byte>.Empty,
                    Connection = connection,
                    ReplyTo = "ignore"
                },
                new NatsJSContext(new NatsConnection())
            },
            null
        )!;

        await msg.AckAsync(new AckOpts { DoubleAck = true });

        await connection
            .Received(1)
            .RequestAsync(
                subject: Arg.Any<string>(),
                data: Arg.Is<ReadOnlySequence<byte>>(b => Encoding.ASCII.GetBytes("+ACK").SequenceEqual(b.ToArray())),
                requestSerializer: Arg.Any<NatsRawSerializer<ReadOnlySequence<byte>>>(),
                replySerializer: Arg.Any<NatsRawSerializer<object>>(),
                cancellationToken: Arg.Any<CancellationToken>()
            );
    }
```

Is there any other way to test this? If not, could we please improve this code path to make it easier to test?

**Comment by mtmk** [comment]
We can make `NatsJSMsg<T>` .ctor public. But I suppose you're after an `INatsJSMsg<T>` interface? see #266

**Reply by paulomf** [reply]
I believe that an interface would surely help with the testing.
Changing the `NatsJSMsg<T>` .ctor to public would still require an instance of NatsJSContext, which in turn needs a NatsConnection type, instead of INatsConnection (harder to test/mock).

**Comment by mtmk** [comment]
btw we have introduced the interface in v2.0.2 ;)

**Reply by paulomf** [reply]
awesome! thank you
